src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+                                 Reference and one Literal
+        :raises GenerationError: If the child Reference is a private variable.
+        :raises GenerationError: If the child Reference is a shared variable.
+        '''
+        from psyclone.psyir.backend.fortran import FortranWriter
+        writer = FortranWriter()
+        # Binary Operation check
+        if node.operator is not \
+           BinaryOperation.ADD and \
+           node.operator is not \
+           BinaryOperation.SUB:
+            raise GenerationError(
+                "Binary Operator of type {0} used "
+                "as in index inside an "
+                "OMPTaskDirective which is not "
+                "supported".format(
+                    node.operator))
+        # We have ADD or SUB BinaryOperation
+        if not( (type(node.children[0]) is \
+           Reference and type(node.children[1]) is\
+           Literal) or (type(node.children[0])\
+           is Literal and type(node.children[1])\
+           is Reference)):
+            raise GenerationError(
+                "Children of BinaryOperation are of "
+                "types {0} and {1}, expected one "
+                "Reference and one Literal when"
+                " used as an index inside an "
+                "OMPTaskDirective.".format(
+                    type(node.children[0]).
+                    __name__, type(node.children[1]).
+                __name__))
+
+        # Have Reference +/- Literal, analyse
+        # and create clause appropriately
+        index_name = "" 
+        if type(node.children[0]) is Reference:
+            index_name = writer(node.children[0].symbol)
+        if type(node.children[1]) is Reference:
+            index_name = writer(node.children[1].symbol)
+        index_private = (index_name in 
+                         parallel_private)
+        if index_private:
+            if index_name in private_list:
+                # FIXME I think this should be
+                # supportable, but need to think
+                if node.operator is \
+                    BinaryOperation.ADD:
+                    index_name = index_name + ("+({0} - {1})"
+                           .format(writer(stop_val)
+                               ,writer(start_val)))
+                else:
+                    index_name = index_name + ("-({0} - {1})"
+                           .format(writer(stop_val)
+                               ,writer(start_val)))
+                index_strings.append(index_name)
+            elif index_name in firstprivate_list:
+                # name +/- stop-start
+                if node.operator is \
+                    BinaryOperation.ADD:
+                    index_name = index_name + ("+({0} - {1})"
+                           .format(writer(stop_val)
+                               ,writer(start_val)))
+                else:
+                    index_name = index_name + ("-({0} - {1})"
```

Comment:
I suspect you should really be building a PSyIR expression here and at line 853.